src/core/slice.c

```diff
@@ -347,6 +348,95 @@ static void slice_enumerate_perpetual(Manager *m) {
                 (void) slice_make_perpetual(m, SPECIAL_SYSTEM_SLICE, NULL);
 }
 
+static int slice_freezer_action_supported_by_children(Unit *s) {
+        int r;
+        _cleanup_closedir_ DIR *d = NULL;
+
+        assert(s);
+
+        r = cg_enumerate_subgroups(SYSTEMD_CGROUP_CONTROLLER, s->cgroup_path, &d);
+        if (r < 0)
+                return r;
+
+        for (;;) {
+                _cleanup_free_ char *subgroup = NULL;
+                Unit *u;
+
+                r = cg_read_subgroup(d, &subgroup);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                u = manager_get_unit(s->manager, subgroup);
+                assert(u);
+
+                if (u->type == UNIT_SLICE) {
+                        r = slice_freezer_action_supported_by_children(u);
+                        if (r <= 0)
+                                return r;
+                } else if (!UNIT_VTABLE(u)->freeze || !UNIT_VTABLE(u)->thaw)
+                        return 0;
+        }
+
+        return 1;
```

Comment:
hmm, if we want to determine the units inside a slice we generally enumerate the units with a UNIT_AFTER dependency and then check if their slice field matches our slice. This works because assigning a unit to a slice implies such a dependency. This is a lot better to check than the actual cgroups structure, because the cgroup structure is effect of our structure but not the structure itself, if you follow what I mean

Reply:
I agree that traversing the internal structures is better than traversing the artifact (e.g. current approach would break on delegated units).
That got me to a question -- is the recursion necessary at all? The kernel behavior is already hierarchical, i.e. it would suffice to freeze the chosen cgroup and update respective unit's descendants based on event notifications.

Reply:
We should track the freeze state in each unit, hence we probably should propagate this down the tree in our code. But you are right, we might not need to trigger this in the kernel on each level, but it should be sufficient to trigger that only in the top slice we operate on and let the kernel propagate this down.